### ISSUE TITLE
Connectivity check in MCS

### ISSUE THREAD

**Issue by kexul** [michellab/BioSimSpace#227]
Recently I was aligning macrocyclic molecules and encountered this error quite frequently: 
```
The merge has changed the molecular connectivity but a ring didn't open/close or change size. Check your atom mapping.
```

I found that setting `sanitize=True` always helped solve this issue in this series of ligands. Just post it here in case someone else might find it helpful. 

Here are the code and input files I've used (would like to provide more input files if needed, got a lot recently...): 
```
ligand2 = BSS.IO.readMolecules(['smile552_ref_smile552~smile953_aligned.inpcrd', 'smile552_AC.prmtop']).getMolecules()[0]
ligand1 = BSS.IO.readMolecules(['smile953_smile552_smile552~smile953_aligned.inpcrd', 'smile953_AC.prmtop']).getMolecules()[0]
try:
    mapping = BSS.Align.matchAtoms(ligand1, ligand2)
    morph = BSS.Align.merge(ligand1, ligand2, mapping, True, True)
except BSS._Exceptions.IncompatibleError as error:
    print(error)
    mapping = BSS.Align.matchAtoms(ligand1, ligand2, sanitize=True)
    morph = BSS.Align.merge(ligand1, ligand2, mapping, True, True)
```
[files.zip](https://github.com/michellab/BioSimSpace/files/7039239/files.zip)

Maybe we could add molecular connectivity checks during the MCS generation procedure?

**Comment by lohedges** [comment]
Thanks for reporting.

Unfortunately full molecular sanitization is beyond the scope of BioSImSpace. We have avoided doing anything where ambiguity is possible and have (mostly) left it up to the user to ensure that their input is sane, e.g. by using existing tools. Where possible we've just tried to catch common errors and report them in a clean way.

The only reason why `sanitize=True` isn't the default option is for consistency with `FESetup` (a pre-cursor to BioSimSpace) and `Lomap`. I agree that it often fixes issues like this and generally advise people to try it as a first test when a mapping fails. However, there have been occasions where sanitization returns a significantly worse mapping, so it's not 100% fail safe.

If there are simple and unambiguous things that we should be checking, then I'd be happy to incorporate these.

**Comment by kexul** [comment]
Thanks for clarifying the philosophy of BioSimSpace.

**Comment by kexul** [comment]
Update: Recently I encountered some particular pairs which `sanitize=True` did not work either, I tried several options and find a workable solution, inspired by this [issue](https://github.com/michellab/BioSimSpace/issues/162): find the MCS use hydrogen removed molecule, then pass it as the prematch in `matchAtoms` function. Usually fallback to sire would happen in these cases, and sire would reconstruct the full MCS with hydrogen.  

Hope it could be helpful for someone who has difficulty matching the atoms.
